src/PhpParser/NodeTraverser/RectorNodeTraverser.php

```diff
@@ -42,10 +49,36 @@ public function refreshPhpRectors(array $rectors): void
     {
         $this->rectors = $rectors;
         $this->visitors = [];
+        $this->visitorsPerNodeClass = [];
 
         $this->areNodeVisitorsPrepared = false;
     }
 
+    /**
+     * We return the list of visitors (rector rules) that can be applied to each node class
+     * This list is cached so that we don't need to continually check if a rule can be applied to a node
+     *
+     * @return NodeVisitor[]
+     */
+    public function getVisitorsForNode(Node $node): array
+    {
+        $nodeClass = $node::class;
+        if (! isset($this->visitorsPerNodeClass[$nodeClass])) {
+            $this->visitorsPerNodeClass[$nodeClass] = [];
+            foreach ($this->visitors as $visitor) {
+                assert($visitor instanceof RectorInterface);
+                foreach ($visitor->getNodeTypes() as $nodeType) {
+                    if (is_a($nodeClass, $nodeType, true)) {
+                        $this->visitorsPerNodeClass[$nodeClass][] = $visitor;
```

Comment:
this code assumes that every rules `public function getNodeTypes()` does only return a non-conditional static array. 

something like (artificial example)

```
    /**
     * @return array<class-string<Node>>
     */
    public function getNodeTypes(): array
    {
        if (phpversion() >= "8") {
           return [New_::class];
        }
        return [ClassLike::class];
    }
```

would break because of the caching.

I don't think its a meaningful BC break, just want to point it out.

Reply:
That example in particular would not be a problem because `phpversion()` returns a single value each time Rector is run. The list of visitors per node type is not cached across runs. But I understand what you mean. In any case I can't think of any case where this function would not return a non-conditional static array